ci: Use commit hash so workflows are not canceled when merging multiple PRs

bthomee opened the pull request:
## High Level Overview of Change

This change changes the CI concurrency group for pushes to the `develop` branch to use the commit hash instead of the target branch.

### Context of Change

When multiple PRs are merged in close succession, where each triggers a CI pipeline upon merge, GitHub will cancel the earlier pipelines due to the concurrency group being the same for each. However, there is value in letting them run independently, since if a failure occurs we will know which PR is to blame. Note that for manually triggered or the nightly scheduled, we can use the existing ref (i.e. `develop`), as we don't expect to ever run more than one such pipeline at the same time.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [X] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

mathbunnyru commented on .github/workflows/on-trigger.yml:
```diff
@@ -50,7 +50,7 @@ on:
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.ref }}
+  group: ${{ github.workflow }}-${{ github.event_name == 'push' && github.sha || github.ref }}
```
We had similar problem in Clio, could you please add a comment here showing why you need it?

Here is our implementation in Clio: XRPLF/clio@develop/.github/workflows/build.yml#L31-L32

bthomee replied:
Thanks, that was very useful. I borrowed one part from the Clio group logic, but kept it otherwise as I have now.
* Merges into the develop branch: unique concurrency group.
* All other cases: shared group based on target branch (=ref).

bthomee committed:
ci: Use commit hash so workflows are not canceled when merging multiple PRs

This change changes the CI concurrency group for pushes to the `develop` branch to use the commit hash instead of the target branch.
